File: src/DotNetPad/DotNetPad.Applications/Samples/NullConditionalOperator.vb
Imports System
Imports System.Collections.Generic

Namespace Waf.DotNetPad.Samples
    Module NullConditionalOperator
        Sub Main()
            ' Use ?. to access the Name property
            Dim person As Person = Nothing
            Console.WriteLine("person?.Name: {0}", If(person?.Name, "Nothing"))
            Dim persons As List(Of Person) = Nothing
			Console.WriteLine("persons?(0).Name: {0}", If(persons?(0).Name, "Nothing"))

            person = New Person With {.Name = "Luke"}
            Console.WriteLine("person?.Name: {0}", If(person?.Name, "Nothing"))
            persons = New List(Of Person) From {person}
            Console.WriteLine("persons?(0).Name: {0}", persons?(0).Name, "Nothing")
        End Sub
    End Module

    Class Person
        Public Property Name As String
    End Class
End Namespace

File: src/DotNetPad/DotNetPad.Applications/Samples/NullConditionalOperator.cs
using System.Runtime.CompilerServices;
using static System.Console;

namespace Waf.DotNetPad.Samples;

internal static class NullConditionalOperator
{
    private static void Main()
    {
        // Use ?. to access the Name property
        Person? person = null;
        WriteLine("person?.Name: {0}", person?.Name ?? "null");
        List<Person?>? persons = null;
        WriteLine("persons?[0]?.Name: {0}", persons?[0]?.Name ?? "null");

        person = new Person() { Name = "Luke" };
        WriteLine("person?.Name: {0}", person?.Name ?? "null");
        persons = new List<Person?>() { person };
        WriteLine("persons?[0]?.Name: {0}", persons?[0]?.Name ?? "null");

        // Use ?. to raise the PropertyChanged event
        WriteLine();
        person!.PropertyChanged += PersonPropertyChanged;
        person.Name = "Han";
    }

    private static void PersonPropertyChanged(object? sender, PropertyChangedEventArgs e)
    {
        if (e.PropertyName == nameof(Person.Name))
        {
            WriteLine("New name: " + ((Person)sender!).Name);
        }
    }
}

public class Person : INotifyPropertyChanged
{
    private string? name;

    public string? Name
    {
        get { return name; }
        set
        {
            if (name == value) return;
            name = value;
            RaisePropertyChanged();
        }
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    protected void RaisePropertyChanged([CallerMemberName] string propertyName = null!)
    {
        OnPropertyChanged(new PropertyChangedEventArgs(propertyName));
    }

    protected virtual void OnPropertyChanged(PropertyChangedEventArgs e)
    {
        PropertyChanged?.Invoke(this, e);
    }
}
